### ISSUE TITLE
Unit tests don't run in Ubuntu

### ISSUE THREAD

**Issue by herter4171** [idaholab/moose#16092]
## Bug Description

One of our users wishes to compile and run individual unit tests for a module.  Our standard container distro is Ubuntu 18.04 with CentOS 8 as a fall-back.  As shown below, what we've found is that unit tests indicate zero tests and zero test cases specifically on Ubuntu.  Everything works in a CentOS container, and we get zero tests and zero test cases on non-container Ubuntu.  From my testing, the problem lies with Ubuntu somehow.

![image](https://user-images.githubusercontent.com/21989132/98299240-3ae5cd80-1f6c-11eb-9dab-8677e3ba2e98.png)

## Steps to Reproduce

Spin up Ubuntu in some fashion, and attempt to build and run a module's unit tests.  @milljm, you'll find more detail in the email chain from a while ago with the subject `Branches and Building Unit Tests in Containers`.

## Impact

I believe we'll need this working for SQA, but it's not preventing much work from being done.

**Comment by WilkAndy** [comment]
See also #15422 , #16083 , and a bit of discussion at https://github.com/idaholab/civet/issues/487

**Comment by herter4171** [comment]
Hey @WilkAndy, I'm glad I'm not the only one.  I think the new thing uncovered here is that it's seemingly distro dependent.

**Comment by herter4171** [comment]
@WilkAndy, I've tested again now that #16217 is merged into master.  To verify that, I ran the following.
```
git log | grep cb3ae6a4c8b6d2cf6322a0cd43611bab1ac6948a
```
Still, when I do the following on both container Ubuntu and non-container Ubuntu, I get the same output as before with zero tests and zero cases.
```
cd $MOOSE_DIR/modules/fluid_properties/unit
make -j `nproc`
./fluid_properties-unit-opt
```
Container CentOS continues to work fine as before.

**Comment by WilkAndy** [comment]
I think that @loganharbour has a fix that means the unit tests will be run by CIVET.  Is that right @loganharbour ?

**Comment by loganharbour** [comment]
@herter4171 when you compile libmesh, does it list that cppunit is available?

**Comment by herter4171** [comment]
On both Ubuntu and CentOS, from grepping `cppunit` during libMesh configuring, I get the following lines.
```
checking for cppunit-config... none
cppunit.......................... : no
```
I see that there are package manager installs.  I'll try an Ubuntu build with `libcppunit-dev` installed to see if anything different happens.

**Comment by loganharbour** [comment]
Sorry @herter4171 didn't see your edit as e-mail notifications aren't sent from edits.

cppunit can be quite pesky and can require some rather specific configuration in some cases. I just went through this on our  build boxes. If libMesh is telling you that you don't have cppunit, it means that it couldn't compile something with cppunit. Which is most likely your issue here.

If you need some assistance on getting cppunit working, ping me tomorrow on Slack.

**Comment by helen-brooks** [comment]
Hi, just wondering what's the current status of this issue? There seem to be a lot of similar issues, some being open and others closed. Is there a fix / PR available, or is it still in progress? I'm also on Ubuntu 20.04, and find I get the lack of finding unit tests in modules, tutorial, and in my own  app, while the ones in unit in the root moose dir run fine. As per this comment https://github.com/idaholab/civet/issues/487#issuecomment-663090153 if I copy-paste the src code and header for the MooseUnitApp and register this instead in the main.C instead of my / as well as/ own app, it will now find the unit tests. I can live with this for now, but just trying to understand if there's a PR in the pipeline.

**Comment by herter4171** [comment]
@helen-brooks, I'm still waiting to hear of a solution, and I'm not aware of any PRs.

**Comment by loganharbour** [comment]
Yeah - I'm not sure what I was thinking about my comments about cppunit and gtest above.

I tried to take a crack at this this weekend one of our test boxes, but hit some login issues. Should be able to take a crack at it tomorrow.

**Comment by loganharbour** [comment]
Whoops. Should not have closed this.

For an update: no success yet after many struggles. We can get unit tests to work correctly on ubuntu when we build the stack with spack. They do _not_ work with Ubuntu or CentOS with our Conda compiler stack.

**Comment by herter4171** [comment]
@loganharbour, thanks for opening this up again, and Happy New Year.  I wasn't going to bust anyones chops about this being closed without seeing behind the curtain, and I'll be interested to see a resolution to pass on to my interested user.

**Comment by loganharbour** [comment]
> I wasn't going to bust anyones chops about this being closed without seeing behind the curtain

Hah - no problem. This is a very frustrating problem that will hopefully have a very satisfying fix 😬

**Comment by hugary1995** [comment]
I'm having the same issue. Just leaving a message here to follow this thread.

**Comment by roystgnr** [comment]
Here's what I think is up:

Google Test uses static initialization as its method of registering tests.  You use its macros, those create some objects, those objects' constructors do the registration before main() begins.

For shared libraries, static initialization happens when the library is dynamically loaded.

On some systems, the runtime dynamic linker skips any libraries it doesn't need to load, and it decides on 'need-to-load' by looking for any symbols that are undefined in the executable (or in already loaded libraries) but are defined in the candidate library; if it finds them, it loads the candidate.

But we have no undefined symbols in main() that are satisfied by the unit test library!  So it never bothers to load, so the initialization doesn't happen, so the unit tests don't get registered.


The workaround was simple: create some function in the library, call it from main().  That takes me from `0 tests from 0 test cases ran.` to `9 tests from 2 test cases ran.` with the stochastic tools unit tests.

Presumably, the undefined-in-main defined-in-library symbol triggers the whole library loading, which triggers static initialization, which registers the tests.

I'm not sure if there's any way to do this that is less of a smelly hack.
